### ISSUE TITLE
TReNA

### ISSUE THREAD

**Issue by marichards** [Bioconductor/Contributions#341]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/PriceLab/TReNA

Confirm the following by editing each check box to '[x]'

- [X] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [X] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [X] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [X] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [X] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [X] The 'devel' branch for new packages and features.
- [X] The stable 'release' branch, made available every six
      months, for bug fixes.
- [X] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by llrs** [comment]
Hi @marichards (Matthew?),

I am interested in your packages because it enables "the construction of gene network models". But I couldn't find more information about how does it creates the gene networks. Maybe you could elaborate more in the vignette about it?

Many thanks!

**Comment by marichards** [comment]
Hi Lluis, 

Thanks for the suggestion, I'll add an example to the vignette to make it more clear. I might also consider adding a more in-depth separate vignette as well. 

Thanks!
Matt

**Comment by vobencha** [comment]
Hi,

The package looks good. Just a few comments below.

1) S4 class slot access

Looks like you have Solvers, Filters and the TReNA class defined. In general, '@' slot access is used when defining getters or setters on S4 classes. Once defined, these functions should be used throughout the code instead of '@'. The rational is that direct slot access can unknowingly corrupt or create an invalid object. Robust getters/setters can check data type and dimension and trigger validity methods when necessary.

i) The getAssayData() getter is defined on the Solver class. This should be used throughout the code instead of obj@mtx.assay. Please replace all of these with getAssayData():
```
~/proj/git/software/TReNA/R >grep 'obj@mtx.assay' *
BayesSpikeSolver.R:      if(rowMeans(obj@mtx.assay)[target.gene] < stats::quantile(rowMeans(obj@mtx.assay), probs = 0.1)){                  
BayesSpikeSolver.R:    mtx <- obj@mtx.assay
EnsembleSolver.R:               if(rowMeans(obj@mtx.assay)[target.gene] < stats::quantile(rowMeans(obj@mtx.assay), probs = 0.1)){                  
EnsembleSolver.R:                   trena <- TReNA(obj@mtx.assay, solver = solver.list[[i]] )
FootprintFilter.R:                  candidate.tfs <- intersect(tbl.out$tf, rownames(obj@mtx.assay))            
...
```

ii)  Same for the getSolverName() defined on the TReNA class. Please replace all of these with getSolverName():
```
~/proj/git/software/TReNA >grep -R 'obj@solver' *
inst/utils/extraFunctions.R:      fit <- trainModel(obj@solver, target.gene, tfs, training.samples, tf.weights)
inst/utils/extraFunctions.R:      prediction <- predictFromModel(obj@solver, model, tfs, test.samples)
R/TReNA.R:      run(obj@solver, target.gene, tfs, tf.weights, extraArgs)
```

2) S4 class man pages

The class man pages should document the methods and any getters/setters that work on that class. At the very least it should mention them in \seealso. For example, the ?FootprintFinder man page does not mention closeDatabaseConnectinos, getGtfGeneBioTypes, getGtfMoleculeTypes, getChromLoc, etc. I realize you have individual man pages for these functions but the problem is how does the user connect the two? How would they know there is a getFootprintsInRegions() function?

i) Please add \detail or \seealso information that helps connect these classes and methods:
```
~/proj/git/software/TReNA/R >grep setMethod *
BayesSpikeSolver.R:setMethod("run", "BayesSpikeSolver",
EnsembleSolver.R:setMethod("run", "EnsembleSolver",
FootprintFilter.R:setMethod("getCandidates", "FootprintFilter",
FootprintFinder.R:setMethod("closeDatabaseConnections", "FootprintFinder",
FootprintFinder.R:setMethod("getGtfGeneBioTypes", "FootprintFinder",
FootprintFinder.R:setMethod("getGtfMoleculeTypes", "FootprintFinder",
FootprintFinder.R:setMethod("getChromLoc", "FootprintFinder",
FootprintFinder.R:setMethod("getGenePromoterRegion", "FootprintFinder",
FootprintFinder.R:setMethod("getFootprintsForGene", "FootprintFinder",
FootprintFinder.R:setMethod("getFootprintsInRegion", "FootprintFinder",
FootprintFinder.R:setMethod("getPromoterRegionsAllGenes","FootprintFinder",
FootprintFinder.R:setMethod("mapMotifsToTFsMergeIntoTable", "FootprintFinder",
LassoPVSolver.R:setMethod("run", "LassoPVSolver",
LassoSolver.R:setMethod("run", "LassoSolver",
NullFilter.R:setMethod("getCandidates", "NullFilter",
PearsonSolver.R:setMethod("run", "PearsonSolver",
RandomForestSolver.R:setMethod("run", "RandomForestSolver",
RidgeSolver.R:setMethod("run", "RidgeSolver",
Solver.R:setMethod("getAssayData", "Solver",
Solver.R:setMethod("rescalePredictorWeights", "Solver",
SpearmanSolver.R:setMethod("run", "SpearmanSolver",
SqrtLassoSolver.R:setMethod("run", "SqrtLassoSolver",
TReNA.R:setMethod("solve", "TReNA",
TReNA.R:setMethod("getSolverName", "TReNA",
VarianceFilter.R:setMethod("getCandidates", "VarianceFilter",
```
ii) Document getAssayData() on the man pages for the classes inheriting from Solver.

Valerie

**Comment by vobencha** [comment]
Also please remove TReNA_Vignette.html from vignettes/.

Valerie

**Comment by marichards** [comment]
@llrs Could you give me a specific example of the sort of network application you'd be interested in? I realize my intended addition to the vignette would be more in the vein of model-based prediction rather than network generation, and I'd like to address your comment as appropriately as possible.

On a related note, we are currently working on interactive graphics to make construction and visualization of such networks more seamless, but this is not part of the current TReNA package. 

Thanks
Matt

**Comment by llrs** [comment]
I am interested in networks and I was exploring packages to build networks and I found TReNA that said it could be used for it. But now that I read the vignette more carefully I realize that is not the purpose of the package. It doesn't create networks as in [WGCNA](https://cran.r-project.org/package=WGCNA) or [DiffCoEx](https://static-content.springer.com/esm/art%3A10.1186%2F1471-2105-11-497/MediaObjects/12859_2010_4080_MOESM1_ESM.DOC) (It is based on WGCNA and it is not build as a package), or [DiffCorr](http://diffcorr.sourceforge.net/).

However, maybe there's the possibility to represent/store the network by which the genes selected with TReNA affect the targeted gene.

**Comment by marichards** [comment]
@vobencha 

Hi Valerie, 

I believe we've addressed all of your comments. With regards to 2ii, we included this documentation in the inherited classes using the /seealso tag and hope this sufficiently covered that comment. We also added two new getters (getSolverObject and getFilterAssayData) to address comment 1. 

Thank you for your comments, they've undoubtedly improved our package.

Thanks,
Matt
